pennylane_lightning/lightning_qubit/_adjoint_jacobian.py

```diff
@@ -0,0 +1,362 @@
+                trainable_params,
+            )
+        jac = np.array(jac)
+        jac = jac.reshape(-1, len(trainable_params))
+        jac_r = np.zeros((jac.shape[0], processed_data["all_params"]))
+        jac_r[:, processed_data["record_tp_rows"]] = jac
+        if hasattr(qml, "active_return"):  # pragma: no cover
+            return self._adjoint_jacobian_processing(jac_r) if qml.active_return() else jac_r
+        return self._adjoint_jacobian_processing(jac_r)
+
+    # pylint: disable=inconsistent-return-statements
+    def calculate_vjp(self, tape: QuantumTape, grad_vec):
+        """Compute the vector-Jacobian products of a tape.
+
+        .. code-block:: python
+
+            statevector = LightningStateVector(num_wires=num_wires)
+            statevector = statevector.get_final_state(tape)
+            vjp = LightningAdjointJacobian(statevector).calculate_vjp(tape, grad_vec)
+
+        computes :math:`\\pmb{w} = (w_1,\\cdots,w_m)` where
+
+        .. math::
+
+            w_k = dy_k \\cdot J_{k,j}
+
+        Here, :math:`dy` is the workflow cotangent (grad_vec), and :math:`J` the Jacobian.
+
+        Args:
+            tape (QuantumTape): Operations and measurements that represent instructions for execution on Lightning.
+            grad_vec (tensor_like): Gradient-output vector, also called dy or cotangent. Must have shape matching the output
+                shape of the corresponding tape, i.e. number of measurements if the return type is expectation.
+
+        Returns:
+            The vector-Jacobian products of a tape.
+        """
+        if tape.shots is not None:
+            warn(
+                "Requested adjoint differentiation to be computed with finite shots. "
+                "The derivative is always exact when using the adjoint differentiation "
+                "method.",
+                UserWarning,
+            )
+
+        measurements = tape.measurements
+        tape_return_type = self._check_adjdiff_supported_measurements(measurements)
+
+        if qml.math.allclose(grad_vec, 0) or tape_return_type is None:
+            return qml.math.convert_like(np.zeros(len(tape.trainable_params)), grad_vec)
+
+        if tape_return_type is State:
+            raise QuantumFunctionError(
+                "Adjoint differentiation does not support State measurements."
+            )
+
+        # Proceed, because tape_return_type is Expectation.
+        if len(grad_vec) != len(measurements):
+            raise ValueError(
+                "Number of observables in the tape must be the same as the "
+                "length of grad_vec in the vjp method"
+            )
+
+        if np.iscomplexobj(grad_vec):
+            raise ValueError(
+                "The vjp method only works with a real-valued grad_vec when the "
+                "tape is returning an expectation value"
+            )
+
+        ham = qml.Hamiltonian(grad_vec, [m.obs for m in measurements])
```

Comment:
We could extend this to a greater variety of measurements (for example vjp with a hamiltonian expectation, with something like:

```suggestion
        ham = qml.simplify(qml.dot(grad_vec, [m.obs for m in measurements]))
```

Reply:
OK.